### ISSUE TITLE
Phase 7c: design leakage-safe held-out K-selection

### ISSUE THREAD

**Issue by H-T-0726** [H-T-0726/expfam-lsm-attributed-relational-data#39]
## Background

Phase 7b / Issue #37 separated three distinct in-sample K-selection diagnostics on the same objective-consistent prototype fits:

- C1 `bic_impl`: selected `k=3` in 6/6 groups (interior)
- C2 `S_cf`: selected `k=7` in 6/6 groups (range boundary)
- C3 `S_laplace_post`: selected `k=3` in 6/6 groups (interior)

The audited Phase 7b result is diagnostic only. It did **not** establish that C1, C2, or C3 is theoretically correct, did not prove K-selection consistency, and did not reproduce paper Experiment 2.

Formal Phase 7b decision: **C — DESIGN_HELDOUT_K_SELECTION_NEXT**.

This issue is the design/audit phase for that decision.

---

## Objective

Design a leakage-safe, K-comparable held-out predictive selection protocol for the attributed relational latent-structure model **before implementing or running any new K-selection experiment**.

The output of this issue is a frozen design and an implementation plan, not a new empirical result.

---

## Non-goals

Do not in this phase:

- run a new K sweep
- claim held-out prediction is theoretically consistent for recovering `K_TRUE`
- call any Q-based score Schwarz BIC
- promote `S_laplace_post` to ELBO / marginal likelihood / corrected BIC
- reproduce paper Experiment 2
- mix old-0.5 / fixed / consistent numerical lineages
- modify manuscript claims

---

## Core design questions

The design must answer all of the following explicitly.

### 1. What exactly is held out?

Consider and distinguish at least:

- **dyad / edge holdout with the same nodes**: transductive Y prediction
- **node-disjoint holdout**: inductive/generalization setting
- **X-entry or X-column holdout**: attribute prediction
- **joint X/Y held-out target**: only if a common leakage-safe definition is defensible

Do not silently treat these as equivalent.

### 2. What is the primary predictive target?

Choose one primary target before any experiment. Candidate examples may include:

- held-out Y mean log score per observed dyad
- held-out X log score / deviance appropriate to the family
- a pre-specified joint score if its weighting and sample unit are defensible

The primary score must be comparable across candidate K and must not use test information in fitting, preprocessing, thresholding, feature construction, node selection, or hyperparameter selection.

### 3. What is the statistical unit?

Specify whether uncertainty / replication is over random splits, nodes, dyads, or datasets. Do not treat dependent dyads as independent replicates merely because there are many pairs.

### 4. How is relational dependence handled?

Random dyad holdout can retain the same nodes in train and test and therefore answers a transductive prediction question. Node-disjoint holdout answers a different, harder question. The design must state which question is primary and why.

### 5. How are latent variables for test observations handled?

For every proposed split, specify whether test-side latent variables are already estimated from training-side observations, inferred from allowed test-side X only, inferred from allowed test-side Y context, or completely new/unavailable. Any use of held-out target values to infer the latent variables used to predict those same targets is leakage and must be prohibited.

### 6. How are X and Y preprocessing steps frozen?

All transformations that can depend on data must be fit on training data only, including where relevant centering/scaling, node/item selection, frequency thresholds, derived attributes, family-specific transforms, and exposure/count summaries. Record provenance for every train-derived object.

### 7. What does K-selection mean here?

Define the operational held-out selector, e.g. `K_hat_pred = argmin/argmax_K held_out_risk(K)`, with exact sign convention and tie rule. Keep this separate from the latent generative `K_TRUE` definition. Predictive-optimal K and generative true K need not coincide under misspecification.

---

## Required design alternatives

Before choosing a final protocol, compare at least these two families of design.

### Design A — transductive dyad holdout

Same node set appears in train and test; only a subset of Y dyads is withheld.

Must specify:

- train mask construction
- test dyad set
- whether X is fully available at fit time
- whether Z estimation can use only train Y + allowed X
- held-out Y score
- split fraction
- minimum train degree / graph connectivity safeguards
- how repeated random splits are interpreted

### Design B — node-disjoint / inductive holdout

Training and test node sets are disjoint or partially disjoint.

Must specify:

- how parameters learned on train nodes transfer to new nodes
- which observations are allowed to infer test-node Z
- what target is predicted after that inference
- whether this is actually supported by the current model/API without changing the estimand

If current implementation cannot make Design B scientifically clean without a larger model change, say so explicitly rather than forcing it.

---

## Leakage checklist

The final design must include a machine-checkable leakage ledger covering at minimum:

- split generated before fitting
- test Y excluded from E-step and M-step
- test target excluded from initialization statistics
- test data excluded from node/item selection
- test data excluded from derived X features
- test data excluded from normalization/scaling
- test data excluded from tuning thresholds
- test data excluded from K-specific preprocessing
- same split reused across all candidate K within a replicate
- data seed independent of K
- model seed recorded separately

If X is used to infer Z for a Y target, document exactly which X is allowed and why that is not target leakage.

---

## K-comparability requirements

Across candidate K within one replicate:

- identical training observations
- identical held-out observations
- identical preprocessing objects
- identical node set where the design requires it
- identical scoring rule
- no K-specific test filtering
- no failed-fit seed rescue

Tie handling must be pre-specified.

---

## Candidate scoring rules

For each supported family, document the predictive score actually available in the current codebase and whether constants omitted from training likelihood matter for **cross-K comparison on the same test data**.

At minimum examine Bernoulli Y log score, Poisson Y log score, and Gaussian X/Y score where applicable.

Do not invent a cross-family weighted joint score unless its weighting is fixed and justified before viewing results.

---

## Relationship to Phase 7b diagnostics

If/when a later pilot is run, the held-out selector should be compared descriptively with C1 `bic_impl`, C2 `S_cf`, and C3 `S_laplace_post`. The held-out target must remain conceptually separate from all three. No criterion should be declared correct solely because it agrees with `K_TRUE` in one synthetic setting.

---

## Theory audit questions

Before implementation, answer:

1. Under the proposed split, what conditional predictive distribution is actually being scored?
2. Does the same observed information condition every candidate K?
3. Is the score proper for the target distribution being predicted?
4. Does latent-variable inference for the test case leak the held-out target?
5. Under model misspecification, are we selecting predictive-optimal K rather than literal `K_TRUE`?
6. Are there identifiability / non-nestedness issues that limit interpretation of K-selection?
7. Is the proposed replication unit valid given dyadic dependence?

Separate these theoretical questions from numerical implementation checks.

---

## Required outputs

Create design-only documentation first, e.g.:

```text
reports/k_selection_theory/
  heldout_k_selection_design_<date>.md
  heldout_k_selection_leakage_matrix_<date>.md
  heldout_k_selection_implementation_plan_<date>.md
```

Do not create result CSV/figures until the design is independently reviewed and approved.

---

## Independent review gate

Before any experiment code is run, perform a read-only review checking at least:

- target and estimand are explicit
- transductive vs inductive question is not mixed
- train/test leakage paths are closed
- same split/data are reused across K
- test target is never used to infer the quantity used to predict itself
- score sign/tie rules are pre-specified
- replication unit is defensible
- predictive-optimal K is not silently equated with `K_TRUE`
- implementation can actually enforce the proposed masks

Any BLOCKER/HIGH finding must be resolved before implementation.

---

## Decision gate after this design phase

Choose exactly one:

- `A: RUN_HELDOUT_K_SELECTION_PILOT`
- `B: IMPLEMENT_REQUIRED_MODEL_SUPPORT_FIRST`
- `C: RETURN_TO_THEORY_BEFORE_PREDICTIVE_SELECTION`
- `D: STOP_K_SELECTION_WORK_FOR_THESIS`

State what evidence would falsify the selected decision.

---

## Acceptance criteria

- [ ] no new EM/K-sweep experiment run in the design phase
- [ ] primary held-out target fixed before implementation
- [ ] transductive vs inductive scope explicit
- [ ] latent-variable inference for held-out cases fully specified
- [ ] leakage matrix completed
- [ ] same-split-across-K rule fixed
- [ ] seed and provenance rules fixed
- [ ] score and tie rule fixed
- [ ] replication/statistical unit fixed
- [ ] predictive-optimal K kept distinct from generative `K_TRUE`
- [ ] independent design audit completed
- [ ] exactly one post-design decision selected

---

## Starting point

Phase 7b merged via PR #38.

Merge commit: `912591ee0992a73deec81c50c697bace53a2e8ec`

Primary Phase 7b report: `reports/k_selection_theory/k_selection_score_pilot_report_20260823.md`

Formal predecessor decision: `C: DESIGN_HELDOUT_K_SELECTION_NEXT`.

Do not edit `main` directly.

**Comment by H-T-0726** [comment]
Phase 7c completed and merged via PR #40.

Merge commit: `2e465395026bf1f5939ab8ef32790c278ec7edb3`

Independent design audit result before final doc fixes: `DESIGN_APPROVED_AFTER_FIX`; after fixes, BLOCKER/HIGH/MEDIUM = 0 and only LOW findings remained.

Frozen primary design:
- transductive dyad holdout
- held-out plug-in Y mean log score
- `PAIR-MASK TOPOLOGY ONLY` split guards
- two-canary complete-run falsification before smoke
- dataset replicate as replication unit
- Design B node-disjoint = not currently supported

Formal decision: `A: RUN_HELDOUT_K_SELECTION_PILOT`, gated by implementation → independent static review → canary falsification → smoke → independent review.

Next: #41 `Phase 7d: implement and falsify held-out K-selection harness`.
